src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs

```diff
@@ -827,6 +844,37 @@ private void InitializePackagesFolderInfo(string packagesFolderPath)
         }
 
         /// <summary>
+        /// Filter down the reinstall list to just the ones we need to reinstall (i.e. the dependencies)
+        /// </summary>
+        private static HashSet<string> GetDependenciesForReinstall(PackageIdentity packageIdentity, IEnumerable<PackageIdentity> newListOfInstalledPackages, IEnumerable<SourcePackageDependencyInfo> available)
+        {
+            var result = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+            CollectDependencies(result, packageIdentity.Id, newListOfInstalledPackages, available);
+            return result;
+        }
+
+        /// <summary>
+        /// A walk through the dependencies to collect the additional package identities that are involved in the current set of packages to be installed
+        /// </summary>
+        private static void CollectDependencies(HashSet<string> result, string id, IEnumerable<PackageIdentity> packages, IEnumerable<SourcePackageDependencyInfo> available)
+        {
+            result.Add(id);
+
+            // we want the exact PackageIdentity for this id
+            PackageIdentity packageIdentity = packages.First(p => p.Id.Equals(id, StringComparison.OrdinalIgnoreCase));
+
+            // now look up the depdnencies of this exact package identity
+            var sourceDepInfo = available
+                .Where(p => PackageComparer.Equals(p, packageIdentity))
+                .SingleOrDefault();
```

Comment:
Also, it's possible the package could be poorly authored and have duplicate dependencies, or in the case of a bad feed it could.